storage/innobase/trx/trx0sys.cc

```diff
@@ -1253,15 +1280,42 @@ ulint
 trx_sys_any_active_transactions(void)
 /*=================================*/
 {
-	ulint	total_trx = 0;
-
 	mutex_enter(&trx_sys->mutex);
 
-	total_trx = UT_LIST_GET_LEN(trx_sys->rw_trx_list)
-		  + UT_LIST_GET_LEN(trx_sys->mysql_trx_list);
+	ulint	total_trx = UT_LIST_GET_LEN(trx_sys->mysql_trx_list);
+
+	if (total_trx == 0) {
```

Comment:
It looks correct: while mysql_trx_list is non-empty, this function returns its length, resulting in a shutdown loop with "Waiting for x active transactions to finish" diagnostics. Once it's empty, it checks rw_trx_list, it's either same length as n_prepared_trx, in which case nothing is done and zero returned (correct), either the active transactions are coverted to prepared ones, which should result in zero return value the same.